### ISSUE TITLE
Is there a way to simulate circular layout?

### ISSUE THREAD

**Issue by gordonwoodhull** [tgdwyer/WebCola#124]
I tried the "wheel" constraints of "Scalable, Versatile and Simple Constrained Graph Layout" along with `linkLength`, but it came out misshapen. 

Admittedly I only gave this a quick try. Will be content with a "no" and a close, just want to be sure I'm not missing an obvious trick here. Thanks!

**Comment by tgdwyer** [comment]
Hi Gordon,

Interesting.  So did you add a custom projection step for these constraints
as per the paper?  A number of iterations are required to get it to form
nice circles.

The better way to do circular constraints is as per this paper:
http://www.researchgate.net/publication/43433414_Layout_with_Circular_and_Other_Non-linear_Constraints_Using_Procrustes_Projection/file/9fcfd511cb7433f8c9.pdf

video:
https://www.youtube.com/watch?v=Q-7CPEhtNRc

Unfortunately it's not implemented in webcola yet though.

Cheers,

Tim


On 13 July 2015 at 17:22, Gordon Woodhull <notifications@github.com> wrote:

> I tried the "wheel" constraints of "Scalable, Versatile and Simple
> Constrained Graph Layout" along with linkLength, but it came out
> misshapen.
>
> Admittedly I only gave this a quick try. Will be content with a "no" and a
> close, just want to be sure I'm not missing an obvious trick here. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgdwyer/WebCola/issues/124>.
>



-- 
Dr Tim Dwyer BSc BCS(Hons) PhD
Senior Lecturer and Larkins Fellow
Caulfield School of Information Technology
Mobile: 0481 240 767
(from overseas: +61 481 240 767)

**Comment by gordonwoodhull** [comment]
I did some more experimenting. It seems to work perfectly well with N>6, just out of order with N=5, and a little squashed with N=6.

![image](https://cloud.githubusercontent.com/assets/1366709/8703003/118bc016-2ae9-11e5-8098-d3bdee89cb07.png)

![image](https://cloud.githubusercontent.com/assets/1366709/8703014/1bedfd1c-2ae9-11e5-88d4-de2d59e7a1d2.png)

![image](https://cloud.githubusercontent.com/assets/1366709/8703030/2b95cad8-2ae9-11e5-83ad-4493d1e5eaec.png)

![image](https://cloud.githubusercontent.com/assets/1366709/8703062/501f44d8-2ae9-11e5-9428-b7d98d07816c.png)

![image](https://cloud.githubusercontent.com/assets/1366709/8703070/5a9e1ab0-2ae9-11e5-9190-8e2125504ab2.png)

I probably need to learn how to do a custom projection step, as you suggest. 

I am guessing you mean that `d3cola.start()` is a high-level wrapper and I need to operate at a lower level in order to control the order the constraints are applied. (This seems like something I might need for more stable hierarchical layouts too.) Or am I off-track completely?

**Comment by gordonwoodhull** [comment]
For anyone else curious, here is my wheel-generation code (going into dc.graph.js):

```
    function gen_node(i) {
        return {
            id: i,
            name: String.fromCharCode(97+i)
        };
    }
    function gen_edge(i, j, length) {
        return {
            source: j,
            target: i,
            sourcename: nodes[j].name,
            targetname: nodes[i].name,
            length: length
        };
    }
 // ...
        var r = 150,
            strutSkip = Math.floor(N/2),
            rimLength = 2 * r * Math.sin(Math.PI / N),
            strutLength = 2 * r * Math.sin(strutSkip * Math.PI / N);
        nodes = new Array(N);
        edges = [];
        for(i = 0; i < N; ++i)
            nodes[i] = gen_node(i);
        for(i = 0; i < N; ++i) {
            edges.push(gen_edge(i, (i+1)%N, rimLength));
            edges.push(gen_edge(i, (i+strutSkip)%N, strutLength));
            if(N%2)
                edges.push(gen_edge(i, (i+N-strutSkip)%N, strutLength));
        }
```

**Comment by timelyportfolio** [comment]
@gordonwoodhull thx for sharing this

**Comment by tgdwyer** [comment]
I see, so are you doing the projection against the fixed distance constraints in the 'tick' handler?

Or have you modified the layout.ts code?  My concern is that some of the lower level code (e.g. descent.ts and rectangle.ts the [projection logic in rectangle.ts](https://github.com/tgdwyer/WebCola/blob/master/WebCola/src/rectangle.ts#L400-553)) cache additional state (such as node positions), so that setting the x and y on nodes from outside of cola.js may not work like you expect - especially if you are also doing overlap removal or other constraints...  but obviously your code is doing something :-).

By the way, if you are serious about circles then the procrustes projection trick really is far superior in terms of speed and robustness and finishes up not being very much code if you can reuse an existing eigen solver [e.g.](http://www.akiti.ca/EigR12Solver.html).

**Comment by gordonwoodhull** [comment]
I'm just specifying `linkLength` - I haven't tried to dig into the cola.js code, although I'd love to contribute in the future. Definitely would not want to modify cola internal values while it's running.

I'm still unsure what you mean by a "custom projection step" - is it using a lower level API than `d3cola.start()`, as I guessed?

**Comment by tgdwyer** [comment]
Ahhh, I see.  In that case you won't get very nice circles if there are any
other attachments to your "wheels".

The project step, as described in "Scalable, Versatile and Simple
Constrained Graph Layout" is where we "project" the node positions for each
fixed-length edge such that the edge length is exactly satisfied.  It is
important to move the nodes as little as possible as described in the
paper.  For simple fixed length edges the projection is pretty trivial.

To do it properly you would need to insert the projection code for this
class of constraints [here](
https://github.com/tgdwyer/WebCola/blob/master/WebCola/src/rectangle.ts#L400-553
).



On 15 July 2015 at 10:46, Gordon Woodhull <notifications@github.com> wrote:

> I'm just specifying linkLength - I haven't tried to dig into the cola.js
> code, although I'd love to contribute in the future. Definitely would not
> want to modify cola internal values while it's running.
>
> I'm still unsure what you mean by a "custom projection step" - is it using
> a lower level API than d3cola.start(), as I guessed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgdwyer/WebCola/issues/124#issuecomment-121690923>.
>



-- 
Dr Tim Dwyer BSc BCS(Hons) PhD
Senior Lecturer and Larkins Fellow
Caulfield School of Information Technology
Mobile: 0481 240 767
(from overseas: +61 481 240 767)

**Comment by gordonwoodhull** [comment]
Yes, it does get distorted when there are other attachments to the wheels. I think it will be good enough for the demo we're putting together, and hopefully I'll have time to contribute the better projections later.

It seems to me that there would be two levels of constraints, much like the axis versus alignment constraints: one for fixed edge lengths, and another to produce "wheel" constraints built on top of fixed-length constraints.
